keps/sig-node/4188-kubelet-pod-readiness-api/README.md

```diff
@@ -0,0 +1,449 @@
+    string podName = 3;
+    bool static = 4;
+	repeated PodCondition conditions = 5;
+    Timestamp DeletionTimestamp = 3;
+}
+
+// PodConditionsResponse returns a stream of List of PodConditionsResponse
+// Whenever a Pod state changes it will return the new list.
+message PodConditionsListResponse {
+	repeated PodInfo Pods = 1;
+}
+
+// PodConditionsRequest contains Pods UID
+message PodConditionsRequest {
+    string podUID = 1;
+}
+```
+
+##### Unit tests
+
+
+##### e2e tests
+
+  - Feature gate enable/disable tests.
+  - Test API work when Kubelet restarts.
+  - Test API work when Control Plane is unreachable.
+
+### Graduation Criteria
+
+#### Alpha
+
+- [ ] Feature implemented behind a feature flag.
+- [ ] e2e tests completed and enabled.
+- [ ] sig-auth input on proposed lack of authorization for this API.
+
+#### Beta
+- [ ] Fix Kubelet restart [issue](https://github.com/kubernetes/kubernetes/issues/100277)
+
+#### GA
+
+- [ ] No major bugs related to usage of API in 3 months cycle.
+
+### Upgrade / Downgrade Strategy
+
+With gRPC the version is part of the service name.
+Old versions and new versions should always be served and listened by the Kubelet.
+
+### Version Skew Strategy
+
+N/A
+
+## Production Readiness Review Questionnaire
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: `PodInfoAPI` is a new feature gate to
+  enable / disable PodInfo API.
+  - Components depending on the feature gate: Kubelet
+
+###### Does enabling the feature change any default behavior?
+
+No.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, through feature gates.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+The API becomes available again. The API is stateless, so no recovery is needed, clients can just consume the data.
+
+###### Are there any tests for feature enablement/disablement?
+
+e2e test will demonstrate that when the feature gate is disabled, the API returns the appropriate error code.
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+Kubelet may fail to start. The new API may cause the Kubelet to crash.
+
+###### What specific metrics should inform a rollback?
+
+`pod_info_endpoint_errors_get` - but only with feature gate `PodInfoAPI` enabled.
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+No.
+
+### Monitoring Requirements
+
+###### How can an operator determine if the feature is in use by workloads?
+
+Look at the `pod_info_endpoint_requests_list` and `pod_resources_info_requests_get` metric exposed by the Kubelet.
```

Comment:
are these the right metrics? did you mean `pod_info` not `pod_resources_info`? Should it be `pod_readiness` not `pod_info`...info seems very generic?

Reply:
fixed